### PR TITLE
Exec stsc add

### PR DESCRIPTION
**Pull request by fierlion**
### Summary
This adds state updates to the exec command manager, and emits a container event for the monitoring of the exec command in the docker task engine.

### Implementation details
Exec Command agent status is tracked in the container.  I added fields to the ExecCommandAgentMetadata, but kept the actual agent status directly in the container itself.  (This might migrate to the Metadata in future, but it breaks a whole lot of tests so I'd prefer to leave it where it is for now.)
Via the execcmd manager, the docker task engine will update the container and metadata state with the `StartedAt` time as well as the reason for any error (`StoppedReason`-> `ManagedAgent.Reason`).  
The docker task engine will emit the container event during its monitoring cycle.

### Testing
Ran All unit tests locally and updated in place.
I'm letting MACIS run the rest and will update as necessary.

New tests cover the changes: yes

### Description for the changelog
Update Execute Command Agent state and emit container events for that state.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/execcmd/manager_start_linux.go [hunk 1]
```diff
@@ -42,7 +44,7 @@ func (m *manager) RestartAgentIfStopped(ctx context.Context, client dockerapi.Do
 		return NotRestarted, nil
 	}
 	execMD := container.GetExecCommandAgentMetadata()
-	if execMD == nil { // This means m.Start has never been invoked
+	if execMD == nil || execMD.StartedAt == nil { // This means m.Start has never been invoked
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:47]: can we encapsulate this condition in a function of `*manager`?

```golang
func (m *manger) isAgentStarted(execMD *container.ExecCommandAgentMetadata) bool {
    return execMD != nil && execMD.StartedAt != nil
}
```

that way we can use `if m.isAgentStarted(execMD)`, and  `if !m.isAgentStarted(execMD)` as appropriate.

File: agent/engine/execcmd/manager_start_linux.go [hunk 2]
```diff
@@ -135,35 +140,45 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
 	}
 	execRes, err := client.CreateContainerExec(ctx, containerId, execCfg, dockerclient.ContainerExecCreateTimeout)
 	if err != nil {
+		execMD.StoppedReason = fmt.Sprintf("create start error: %v", err)
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:143]: hmm, in these case the StoppedReason will be lost because `SetExecCommandAgentMetadata` is never invoked. I'm thinking invoking `SetExecCommandAgentMetadata` and  `SetKnownExecCommandAgentStatus` should be invoked in `m.StartAgent` rather than here.

File: agent/engine/execcmd/manager_start_linux.go [hunk 3]
```diff
@@ -135,35 +140,45 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
...
-		DockerExecID: execRes.ID,
-		CMD:          execAgentCmd,
-	})
+	nowTime := time.Now()
+	execMD.PID = strconv.Itoa(inspect.Pid)
+	execMD.DockerExecID = execRes.ID
+	execMD.CMD = execAgentCmd
+	execMD.StartedAt = &nowTime
+	container.SetExecCommandAgentMetadata(execMD)
+	container.SetKnownExecCommandAgentStatus(apicontainerstatus.ManagedAgentRunning)
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:182]: As mentioned before, it's probably better if we do this one level up in `m.StartAgent`

File: agent/engine/docker_task_engine_test.go [hunk 4]
```diff
@@ -3174,8 +3198,43 @@ func TestMonitorExecAgentRunning(t *testing.T) {
...
+		}
+		if tc.expectContainerEvent {
+			go func(stateChangeEvents <-chan statechange.Event, expectedManagedAgent apicontainer.ManagedAgent) {
+				for event := range stateChangeEvents {
+					t.Logf("event: %v", event)
+					if containerEvent, ok := event.(api.ContainerStateChange); ok {
+						if containerEvent.ManagedAgents != nil {
+							// there is currently only ever a single managed agent
+							assert.Equal(t, expectedManagedAgent.Status, containerEvent.ManagedAgents[0].Status)
+							t.Log("asserted")
```
Comment (sharanyad) [inline agent/engine/docker_task_engine_test.go:3215]: leftover from debugging?

File: agent/engine/execcmd/manager_start_linux.go (hunk 1 above)
Comment (sharanyad) [reply agent/engine/execcmd/manager_start_linux.go:47]: lets add read lock as well

File: agent/engine/execcmd/manager_start_linux.go [hunk 5]
```diff
@@ -135,35 +140,49 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
 	}
 	execRes, err := client.CreateContainerExec(ctx, containerId, execCfg, dockerclient.ContainerExecCreateTimeout)
 	if err != nil {
+		execMD.StoppedReason = fmt.Sprintf("create start error: %v", err)
```
Comment (sharanyad) [inline agent/engine/execcmd/manager_start_linux.go:143]: can we add more customer-friendly stopped reason since this is directly used by customers? same comment for other places too.

File: agent/engine/docker_task_engine_test.go (hunk 4 above)
Comment (fierlion) [reply agent/engine/docker_task_engine_test.go:3215]: sorry I pushed my changes too early -- they weren't ready for review yet.

File: agent/engine/docker_task_engine_test.go [hunk 6]
```diff
@@ -3174,8 +3190,20 @@ func TestMonitorExecAgentRunning(t *testing.T) {
...
-		// TODO: [ecs-exec] Add validation for calls to Control Plane if agent was restarted
+		taskEngine.(*DockerTaskEngine).monitorExecAgentRunning(ctx, mTestTask, testTask.Containers[0])
+
+		expectedManagedAgent := apicontainer.ManagedAgent{
+			Status:        apicontainerstatus.ManagedAgentRunning,
+			Name:          "EXECUTE_COMMAND_AGENT",
+			LastStartedAt: &nowTime,
+		}
+
+		go func(stateChangeEvents <-chan statechange.Event, expectedManagedAgent apicontainer.ManagedAgent) {
```
[the remark this answers is not in the document]
Comment (fierlion) [reply agent/engine/docker_task_engine_test.go]: updated (please resolve if happy)

File: agent/engine/docker_task_engine_test.go [hunk 7]
```diff
@@ -3213,6 +3250,19 @@ func TestMonitorExecAgentProcesses(t *testing.T) {
...
 	<-restartCtx.Done()
 	time.Sleep(5 * time.Millisecond)
+
+	expectedManagedAgent := apicontainer.ManagedAgent{
+		Status:        apicontainerstatus.ManagedAgentRunning,
+		Name:          "EXECUTE_COMMAND_AGENT",
+		LastStartedAt: &nowTime,
+	}
+
+	go func(stateChangeEvents <-chan statechange.Event, expectedManagedAgent apicontainer.ManagedAgent) {
```
[the remark this answers is not in the document]
Comment (fierlion) [reply agent/engine/docker_task_engine_test.go]: updated (please resolve if happy)

File: agent/engine/docker_task_engine.go [hunk 8]
```diff
@@ -302,9 +303,9 @@ func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
 	_, err = engine.execCmdMgr.RestartAgentIfStopped(ctx, engine.client, task, c, dockerID)
 	if err != nil {
 		seelog.Errorf("Task engine [%s]: Failed to restart ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
-		// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
 	}
-	// TODO: [ecs-exec] call control plane to report ExecCommandAgent RESTARTED/UNKNOWN? if RestartIfStopped returns true
+	seelog.Infof("restarted")
+	mTask.emitContainerEvent(mTask.Task, c, "")
```
Comment (angelcar) [inline agent/engine/docker_task_engine.go:308]: This should be done only if the result of `RestartAgentIfStopped` is  `execcmd.Restarted`.
See https://github.com/aws/amazon-ecs-agent/blob/feature/ecs_exec/agent/engine/execcmd/manager.go#L47

File: agent/engine/docker_task_engine_test.go [hunk 9]
```diff
@@ -3174,33 +3197,61 @@ func TestMonitorExecAgentRunning(t *testing.T) {
...
 	ctx, cancel := context.WithCancel(context.TODO())
 	defer cancel()
 	ctrl, _, _, taskEngine, _, _, _ := mocks(t, ctx, &defaultConfig)
+	nowTime := time.Now()
 	dockerTaskEngine := taskEngine.(*DockerTaskEngine)
 	execCmdMgr := mock_execcmdagent.NewMockManager(ctrl)
 	dockerTaskEngine.execCmdMgr = execCmdMgr
 	defer ctrl.Finish()
+	stateChangeEvents := taskEngine.StateChangeEvents()
+	//defer discardEvents(stateChangeEvents)()
```
Comment (angelcar) [inline agent/engine/docker_task_engine_test.go:3230]: should we remove this?

File: agent/engine/execcmd/manager_start_linux.go [hunk 10]
```diff
@@ -42,7 +44,9 @@ func (m *manager) RestartAgentIfStopped(ctx context.Context, client dockerapi.Do
 		return NotRestarted, nil
 	}
 	execMD := container.GetExecCommandAgentMetadata()
-	if execMD == nil { // This means m.Start has never been invoked
+	if !m.isAgentStarted(execMD) {
+		//if execMD == nil || execMD.StartedAt == nil { // This means m.Start has never been invoked
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:48]: should we remove this?

File: agent/engine/execcmd/manager_start_linux.go [hunk 11]
```diff
@@ -42,7 +44,9 @@ func (m *manager) RestartAgentIfStopped(ctx context.Context, client dockerapi.Do
...
+		seelog.Infof("exec md: %v", execMD)
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:49]: This log should probably be removed? or made Debug and more descriptive?

File: agent/engine/execcmd/manager_start_linux.go [hunk 12]
```diff
@@ -135,35 +143,44 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
...
-		PID:          strconv.Itoa(inspect.Pid),
-		DockerExecID: execRes.ID,
-		CMD:          execAgentCmd,
-	})
-	return nil
+	nowTime := time.Now()
+	execMD.PID = strconv.Itoa(inspect.Pid)
+	execMD.DockerExecID = execRes.ID
+	execMD.CMD = execAgentCmd
+	execMD.StartedAt = nowTime
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux.go:183]: nit: You can use `time.Now()` directly here and get rid of the `nowTime` variable

File: agent/engine/execcmd/manager_start_linux_test.go [hunk 13]
```diff
@@ -160,13 +166,14 @@ func TestStartAgent(t *testing.T) {
...
-				assert.Equal(t, &apicontainer.ExecCommandAgentMetadata{
-					PID:          strconv.Itoa(testPid1),
-					DockerExecID: testDockerExecId,
-					CMD:          filepath.Join(ContainerBinDir, BinName),
-				}, testTask.Containers[0].GetExecCommandAgentMetadata())
+				execMD := testTask.Containers[0].GetExecCommandAgentMetadata()
+				assert.Equal(t, strconv.Itoa(testPid1), execMD.PID, "PID not equal")
+				assert.Equal(t, testDockerExecId, execMD.DockerExecID, "DockerExecId not equal")
+				assert.Equal(t, filepath.Join(ContainerBinDir, BinName), execMD.CMD)
+				assert.Equal(t, test.expectedStatus, execMD.Status, "Exec status not equal")
```
Comment (angelcar) [inline agent/engine/execcmd/manager_start_linux_test.go:176]: can we assert here that `StartedAt` is _zero value_ for the cases where we expect `expectedStatus == apicontainerstatus.ManagedAgentStopped` and _non zero value_ for `expectedStatus == apicontainerstatus.ManagedAgentRunning`?

File: agent/engine/docker_task_engine.go (hunk 8 above)
Comment (fierlion) [reply agent/engine/docker_task_engine.go:308]: We need to emit an event if we succeed or fail to restart.  (succeed means STOPPED->RUNNING or RUNNING->RUNNING while fail means RUNNING->STOPPED or STOPPED-> STOPPED.)  
I'm adding a comment to clarify:
+       // whether we restarted or failed to restart, we'll want to emit a state change event
+       // redundant state change events like RUNNING->RUNNING are allowed

File: agent/engine/execcmd/manager_start_linux_test.go (hunk 13 above)
Comment (fierlion) [reply agent/engine/execcmd/manager_start_linux_test.go:176]: sure.  I had similar time test cases in `TestGetKnownManagedAgents`.  I'll update this to be in line with those test cases.

File: agent/engine/docker_task_engine.go [hunk 14]
```diff
@@ -302,9 +303,11 @@ func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
 	_, err = engine.execCmdMgr.RestartAgentIfStopped(ctx, engine.client, task, c, dockerID)
 	if err != nil {
 		seelog.Errorf("Task engine [%s]: Failed to restart ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
-		// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
 	}
-	// TODO: [ecs-exec] call control plane to report ExecCommandAgent RESTARTED/UNKNOWN? if RestartIfStopped returns true
+	seelog.Infof("restarted")
+	// whether we restarted or failed to restart, we'll want to emit a state change event
+	// redundant state change events like RUNNING->RUNNING are allowed
+	mTask.emitContainerEvent(mTask.Task, c, "")
```
Comment (angelcar) [inline agent/engine/docker_task_engine.go:310]: I will approve this, but still have my reservations about it because we should be able to only send the event when necessary. For use cases with lots of containers, we probably don't want to send hundreds or possibly thousands of redundant events

**Review by angelcar — APPROVED** [review]

File: agent/engine/docker_task_engine.go [hunk 15]
```diff
@@ -302,9 +303,11 @@ func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
 	_, err = engine.execCmdMgr.RestartAgentIfStopped(ctx, engine.client, task, c, dockerID)
 	if err != nil {
 		seelog.Errorf("Task engine [%s]: Failed to restart ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
-		// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
 	}
-	// TODO: [ecs-exec] call control plane to report ExecCommandAgent RESTARTED/UNKNOWN? if RestartIfStopped returns true
+	seelog.Infof("restarted")
```
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:307]: nit: remove this

File: agent/engine/docker_task_engine.go (hunk 14 above)
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:310]: why is this needed in case of a successful restart?

Comment (sharanyad) [inline agent/engine/docker_task_engine.go:310]: could we add a third argument (reason)?

File: agent/engine/execcmd/manager_start_linux.go [hunk 16]
```diff
@@ -135,35 +141,43 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
...
+		execMD.StoppedReason = fmt.Sprintf("Unable to start Execute Command Agent [pre-start]: %v", err)
+		execMD.Status = apicontainerstatus.ManagedAgentStopped
+		return execMD, StartError{error: err, retryable: true}
 	}
 	seelog.Debugf("Task engine [%s]: sent ExecCommandAgent start signal for container: %s ->  docker exec id: %s", task.Arn, containerId, execRes.ID)
 
 	inspect, err := client.InspectContainerExec(ctx, execRes.ID, dockerclient.ContainerExecInspectTimeout)
 	if err != nil {
-		return StartError{error: err, retryable: true}
+		execMD.StoppedReason = fmt.Sprintf("Unable to start Execute Command Agent [docker inspect]: %v", err)
```
Comment (sharanyad) [inline agent/engine/execcmd/manager_start_linux.go:161]: `docker inspect` is not very helpful for customers. would suggest removing it.

File: agent/engine/execcmd/manager_start_linux.go [hunk 17]
```diff
@@ -135,35 +141,43 @@ func (m *manager) doStartAgent(ctx context.Context, client dockerapi.DockerClien
...
-		DockerExecID: execRes.ID,
-		CMD:          execAgentCmd,
-	})
-	return nil
+	execMD.PID = strconv.Itoa(inspect.Pid)
+	execMD.DockerExecID = execRes.ID
+	execMD.CMD = execAgentCmd
+	execMD.StartedAt = time.Now()
+	execMD.Status = apicontainerstatus.ManagedAgentRunning
+	return execMD, nil
```
Comment (sharanyad) [inline agent/engine/execcmd/manager_start_linux.go:182]: shouldn't we be emitting a container event for successful agent start here?

File: agent/engine/docker_task_engine.go (hunk 14 above)
Comment (fierlion) [reply agent/engine/docker_task_engine.go:310]: we need to emit a container event for either our failure or our success cases here.
If the agent was in STOPPED state, then this would emit RUNNING STSC (which would mean STOPPED -> RUNNING)

Comment (fierlion) [reply agent/engine/docker_task_engine.go:310]: the container event itself has the reason embedded in the `ExecCommandAgentMetadata.Reason`

File: agent/engine/execcmd/manager_start_linux.go (hunk 16 above)
Comment (fierlion) [reply agent/engine/execcmd/manager_start_linux.go:161]: for debug, this tells us where the failure occurred:

`Unable to start Execute Command Agent [create]`
`Unable to start Execute Command Agent [pre-start]`
`Unable to start Execute Command Agent [docker inspect]`

would it be ok to just call it `[inspect]` to differentiate for debug purposes?

Comment (sharanyad) [reply agent/engine/execcmd/manager_start_linux.go:161]: sgtm

File: agent/engine/docker_task_engine.go (hunk 14 above)
Comment (sharanyad) [reply agent/engine/docker_task_engine.go:310]: ah ok, adding a comment will clarify :)

File: agent/engine/execcmd/manager_start_linux.go (hunk 17 above)
Comment (fierlion) [reply agent/engine/execcmd/manager_start_linux.go:182]: It's emitted by the caller(s) of this function in agent/engine/docker_task_engine.go (see comment above)

Comment (sharanyad) [reply agent/engine/execcmd/manager_start_linux.go:182]: my bad, I was thinking about emitting container event when we start exec agent for the first time:
https://github.com/aws/amazon-ecs-agent/blob/feature/ecs_exec/agent/engine/docker_task_engine.go#L1267-L1272 

are we planning to add this in a different PR?

Comment (fierlion) [reply agent/engine/execcmd/manager_start_linux.go:182]: Added it.  (This was a pretty major miss on my part.)

File: agent/engine/docker_task_engine.go [hunk 18]
```diff
@@ -1261,9 +1263,15 @@ func (engine *DockerTaskEngine) startContainer(task *apitask.Task, container *ap
...
 		if err := engine.execCmdMgr.StartAgent(engine.ctx, engine.client, task, container, dockerID); err != nil {
 			seelog.Errorf("Task engine [%s]: Failed to start ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
-			// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
 		}
-		// TODO: [ecs-exec] call control plane to report ExecCommandAgent SUCCESS here
+		// whether we started or failed to start, we'll want to emit a state change event
+		// redundant state change events like RUNNING->RUNNING are allowed
+		mTask, ok := engine.managedTasks[task.Arn]
+		if ok {
+			mTask.emitContainerEvent(mTask.Task, container, "")
```
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:1271]: missing unit tests for this

File: agent/engine/docker_task_engine.go [hunk 19]
```diff
@@ -1261,9 +1263,15 @@ func (engine *DockerTaskEngine) startContainer(task *apitask.Task, container *ap
...
+		} else {
+			seelog.Errorf("Task engine [%s]: Failed to update status of ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
```
Comment (sharanyad) [inline agent/engine/docker_task_engine.go:1273]: logging the container name might be more readable here.

**Review by sharanyad — APPROVED** [review]

File: agent/engine/docker_task_engine.go [hunk 20]
```diff
@@ -1266,10 +1268,16 @@ func (engine *DockerTaskEngine) startContainer(task *apitask.Task, container *ap
...
-			// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
+			seelog.Errorf("Task engine [%s]: Failed to start ExecCommandAgent Process for container [%s]: %v", task.Arn, container.Name, err)
+		}
+		// whether we started or failed to start, we'll want to emit a state change event
+		// redundant state change events like RUNNING->RUNNING are allowed
+		mTask, ok := engine.managedTasks[task.Arn]
+		if ok {
+			mTask.emitContainerEvent(mTask.Task, container, "")
+		} else {
+			seelog.Errorf("Task engine [%s]: Failed to update status of ExecCommandAgent Process for container [%s]: %v", task.Arn, container.Name, err)
```
Comment (angelcar) [inline agent/engine/docker_task_engine.go:1279]: I think there will be no `err` here. Maybe this should say just `seelog.Errorf("Task engine [%s]: Failed to update status of ExecCommandAgent Process for container [%s]: managed task not found", task.Arn, container.Name)`

**Review by angelcar — APPROVED** [review]

File: agent/engine/docker_task_engine.go (hunk 20 above)
Comment (fierlion) [reply agent/engine/docker_task_engine.go:1279]: I'll change this with my next PR to add in the updated model.
